### PR TITLE
kconfig: kconfiglib.py: Backup files only

### LINKED ISSUE #31362: kconfiglib.py _save_old() may rename /dev/null -- replacing /dev/null with a file
**Describe the bug**

the kconfiglib.py _save_old("/dev/null") may actually, unexpectly, succeed. 

afterwards /dev/null was opened for writing as a (new) file 

**To Reproduce**
1. mkdir build; cd build
2. cmake -DBOARD=board\_xyz
3. ls -l /dev/null
4. cat /dev/null

**Expected behavior**

/dev/null is still device node

**Impact**

/dev/null is a file, with kconfig content

**What impact does this issue have on your progress**

host code that opened /dev/null as config file got "garbage"

**Environment (please complete the following information):**

- a docker container (not as root, but probably enough capabilites to rename /dev/null)

zephyr v1.14 -- but after investigating main in 2021-01 it still can happen

** how to fix **

`if not os.path.isfile(path) return` to skip rename if path does not point to a file
(either directly, or via symlink)

### PR DESCRIPTION
**Pull request by domo141**
The _save_old() to return early if <filename> is not file (or symlink
to a file).

This is simplest alternative to avoid attempt to rename /dev/null
(which could succeed).

This also keeps fifos (perhaps nonexistent potential usage but
this is nicer).

If <filename> were directory or socket, after shutil.copyfile(),
writing to the file (by caller, _write_config()), would fail.

Fixes #31362

Signed-off-by: Tomi Ollila <tomi.ollila@iki.fi>

### CODE REVIEW

**Review by nashif — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
I can't reproduce this issue.

**Review by carlescufi — CHANGES_REQUESTED** [review]
Before we consider merging this, could you please provide instructions on how to reproduce this step by step in order to verify that the issue is reproducible in other systems?

**Review by tejlmand — APPROVED** [review]
Haven't reproduced, but it seems sane to me that we only try to copy something if it already exists on the file system.

**Comment by domo141** [comment]
Reproducing in normal systems might be possible by making /dev/ world writable, or run commands as root
(and I don't recommend trying).

In my case, in zephyr 1.14.2 based tree, in our build system that has happened since middle of 2020
-- we noticed this after the build we started to run `dnsmasq -C /dev/null ...` which failed due to "garbage"
in the config file /dev/null (kconfig output)

excerpts from my log while finding where /dev/null changes from device node to regular file:
```
+ ls -l /dev/null
crw-rw-rw-. 1 root root 1, 3 Jan 15 12:55 /dev/null

+ strace -f cmake ..
 ...
[pid    76] lstat("/dev/null", {st_mode=S_IFCHR|0666, st_rdev=makedev(1, 3), ...}) = 0
[pid    76] rename("/dev/null", "/dev/.null.old") = 0
[pid    76] open("/dev/null", O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0666) = 3
[pid    76] fstat(3, {st_mode=S_IFREG|0644, st_size=0, ...}) = 0
...
exit_group(0)                           = ?
+++ exited with 0 +++

+ ls -l /dev/null
-rw-r--r--. 1 root root 12547 Jan 15 12:56 /dev/null
+ cat /dev/null
# Generated by Kconfiglib (https://github.com/ulfalizer/Kconfiglib)
...
``` 
This is was run on zephyr 1.14.2 tree -- but can happen in current HEAD tree, renamed file name
would be a bit different (/dev/null.old)

In our case build was done in jenkins jnlp docker container, user was not root.

File: scripts/kconfig/kconfiglib.py
Comment (mbolivar-nordic) [inline scripts/kconfig/kconfiglib.py:1548]: Maybe this:

```suggestion
          due to permissions errors.)
```

It's not true that errors are ignored if the file is a directory, because `os.path.isfile` returns `False` for directories (at least on my computer), so the attempt to write a directory is never made.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
The `strace` output is convincing.

I think one of the comments is not quite accurate, but that's not a blocker.

This is a copy of kconfiglib, though; could you please open up a pull request in https://github.com/ulfalizer/Kconfiglib/ and give a link to it from the commit log here?

I don't want to lose track of this in the mainline repository if Ulf resurfaces.

File: scripts/kconfig/kconfiglib.py
Comment (domo141) [reply scripts/kconfig/kconfiglib.py:1548]: Here we're referring <filename>.old, not <filename> which was checked by os.path.isfile() in my change.

Anyway, this change you suggested could be less confusing, and therefore could be better of what
I changed in that commit.

Should I do the pull request now, or wait for this to get its final form, so the pull request could
have exactly same commit message contents and exactly same change?

Comment (domo141) [reply scripts/kconfig/kconfiglib.py:1548]: hmm, forget my above comment about `<filename>.old` and `<filename>` ... I must have been too
tired to examine the content well enough. The other issue (kconfig pull request) got my attention there...

Comment (domo141) [reply scripts/kconfig/kconfiglib.py:1548]: Ok, The initial changes in comments were just to do minimal potentially good enough change to
the related comments. The changes in line 1548 was made to match the change in comment lines
in 6420 (6424 after change). 

Now I looked whether the above change could be done w/o the change in 6424 (or should that be
changed also). 

To me it looks like the change in 1548 can be made w/ change in 6424.

I almost pressed the 'Commit suggestion' button, but then noticed that perhaps `s/.)/)./` there (i.e.
trailing perions after closing parentheses). Perhaps I accept, then re-edit (if necessary)

**Comment by domo141** [comment]
Ok, not check fails due to my edits from github interface.

How should I fix this in a proper way for this repository?

(force-push new change to my branch where I pick commit message change from nashif and 
 add Co-authored-by: line from one of the commits?)

**Comment by tejlmand** [comment]
> Ok, not check fails due to my edits from github interface.

Yep, this is one of the limitation of the github UI.
It's necessary to do a minor cleanup afterwards in the PR, but that is easy though.

> 
> How should I fix this in a proper way for this repository?
> 

Just do a interactive rebase and squash the commits together into one.
```
git rebase -i 79d1e0605f8a131aea6c517445b62ea5e9721900~3
```
and then squash the last two commits into your original one, and then forcepush your branch.

**Comment by domo141** [comment]
Thank you @tejlmand, now I know :)

**Comment by mbolivar-nordic** [comment]
Thanks for adjusting the comment!

> could you please open up a pull request in https://github.com/ulfalizer/Kconfiglib/ and give a link to it from the commit log here?

Do you plan on doing this? I think we should not merge this PR without that done.

**Comment by domo141** [comment]
PR opened. sequence: git format-patch, emacs, cd, git am, git commit --amend, git push origin HEAD:topic...

**Comment by mbolivar-nordic** [comment]
> PR opened. 

Thanks!

> sequence: git format-patch, emacs, cd, git am, git commit --amend, git push origin HEAD:topic...

For future reference if you prefer to use the web interface: you can also just go to https://github.com/ulfalizer/Kconfiglib/pulls, click "New pull request", click "compare across forks", and choose "domo141/Kconfiglib" and "topic-devnull-40bb230" in the drop down menus, then create the PR.

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
@carlescufi please revisit. I think this should go into 2.5 since it fixes a bug in containerized environments.

**Review by carlescufi — APPROVED** [review]
Thanks for the explanations

**Pull request merged**
